src/noob/scheduler.py

```diff
@@ -0,0 +1,130 @@
+from graphlib import TopologicalSorter
+from itertools import count
+
+from pydantic import BaseModel, PrivateAttr
+
+from noob.event import Event
+from noob.node import Edge, Node
+
+
+class NodeCoord(BaseModel):
+    """
+    Coordinates of the nodes expressed with epoch and node ID.
+
+    graphlib.TopologicalSorter can only "ready" node_ids of the same epoch.
+    So, we add a way to track which epoch of the node_id is ready.
+
+    """
+
+    epoch: int
+    id: str
+
+
+class Scheduler(BaseModel):
+    """
+    Extends the functionality of :class:`.TopologicalSorter`, by wrapping
+    a set of epoch-assigned :class:`.TopologicalSorter`, so that each graph
+    (and the nodes within) can be linked to an epoch.
+
+    """
+
+    _clock: count = PrivateAttr(default_factory=count)
+    _graphs: dict[int, TopologicalSorter] = PrivateAttr(default_factory=dict)
+
+    def update(self, events: list[Event]) -> None:
+        """
+        When a set of events are received, update the graphs within the scheduler.
+
+        """
+        if not events:
+            return
+
+        for event in events:
+            self.done(epoch=event["epoch"], node_id=event["node_id"])
+
+    def is_active(self, epoch: int | None = None) -> bool:
+        """
+        Scheduler remains active while it holds at least one graph that is active.
+
+        """
+        if epoch is not None:
+            return self._graphs[epoch].is_active()
+        else:
+            return any(graph.is_active() for graph in self._graphs.values())
+
+    def get_ready(self, epoch: int | None = None) -> list[NodeCoord]:
+        """
+        Output the set of nodes that are ready across different graphs and epochs.
+
+        """
+        graphs = self._graphs.items() if epoch is None else [(epoch, self._graphs[epoch])]
+
+        ready_nodes = [
+            NodeCoord(epoch=e, id=node_id) for e, graph in graphs for node_id in graph.get_ready()
+        ]
+
+        return ready_nodes
+
+    def done(self, epoch: int, node_id: str) -> None:
+        self._graphs[epoch].done(node_id)
+
+    def evict_cache(self) -> Event:
+        """
+        We can evict the cached event from the node once all successors
+        are marked "done."
+
+        Not implemented in :class:`.Scheduler` while we're still utilizing
+        :class:`.TopologicalSorter`, since its API access to successors /
+        predecessors is limited.
+
+        """
+        raise NotImplementedError()
+
+    def add_graph(self, nodes: dict[str, Node], edges: list[Edge]) -> None:
```

Comment:
it's a *little* awkward to me to need to pass the nodes each time, i would expect `add_graph` to mean "i add a set of nodes and edges, and now that i know about the structure of the graph, i can make `TopologicalSorters` from that. i get why this would be good to not make the scheduler need to have `disable/enable` and other methods, but it seems like it will result in a lot of repeated operations.